corpus_analysis/document.py

```diff
@@ -637,3 +640,83 @@ def update_metadata(self, new_metadata):
+        char_dict = {lab: {} for lab in labels.split(',')}
+        for ch in labels_char:
+            cat = char_dict[ch[0]]
+            cat[ch[1]] = cat.get(ch[1], 0) + 1
+        people = char_dict['PERSON']
+        people_sorted = [(p, people[p]) for p in people if p not in HONORIFICS]
+        people_sorted = sorted(people_sorted, key=lambda p: p[1], reverse=True)
+        cutoff = len(people_sorted)
+        for i in range(len(people_sorted)):
+            if people_sorted[i][1] < cutoff_num:
+                cutoff = i
+                break
+        char_list = people_sorted[:cutoff]
+
+        return char_list
+
+    @staticmethod
+    def filter_honr(name):
+        name = name.split(' ')
+        return [n for n in name if n not in HONORIFICS]
+
+    def char_name_disambiguation(self, char_list):
+        """given a list of char names in a document, group them by potential nicknames
+        :param: a list of character as well as their freq from get_char_list
+        :return: a list of list of character names and freq where the first one is the name,
+        followed by nicknames
+        >>> from corpus_analysis import document
+        >>> from pathlib import Path
+        >>> from gender_analysis import common
+        >>> document_metadata = {'author': 'Austen, Jane', 'title': 'Persuasion', 'date': '1818',
+        ...                      'filename': 'austen_persuasion.txt',
+        ...                      'filepath': Path(common.TEST_DATA_PATH, 'sample_novels',
+        ...                                       'texts', 'austen_persuasion.txt')}
+        >>> persuasion = document.Document(document_metadata)
+        >>> persuasion_chars = persuasion.get_char_list(20)
+        >>> disamb = persuasion.char_name_disambiguation(persuasion_chars)
+        >>> disamb
+        [[('Anne', 425)], [('Captain Wentworth', 119), ('Captain Benwick', 32), ('Wentworth', 31), ('Captain Harville', 21)], [('Lady Russell', 116)], [('Charles', 115), ('Charles Hayter', 29)], [('Mary', 113)], [('Sir Walter', 95)], [('Elizabeth', 82)], [('Elliot', 76), ('Miss Elliot', 32)], [('Louisa', 66)], [('Henrietta', 65)], [('Mrs Musgrove', 40), ('Musgrove', 24)], [('Mrs Smith', 36)], [('Mrs Clay', 33)], [('Miss Elliot', 32)], [('Captain Benwick', 32), ('Benwick', 26), ('Captain Harville', 21)], [('Wentworth', 31)], [('Charles Hayter', 29)], [('Mrs Croft', 27)], [('Benwick', 26)], [('Musgrove', 24)], [('Uppercross', 23)], [('Lady Dalrymple', 22)]]
+        >>> len(disamb)
+        22
+        """
+        to_return = []
+        for i in range(len(char_list) - 1):
+            char_cluster = [char_list[i]]
+            for j in range(i + 1, len(char_list)):
+                if set(self.filter_honr(char_list[i][0])).intersection(
+                set(self.filter_honr(char_list[j][0]))):
+                    char_cluster.append(char_list[j])
+            to_return.append(char_cluster)
+        return to_return
```

Comment:
As we talked about in Slack briefly, this method probably requires some thinking through. If I'm reading the test output in 710 correctly, it looks like the disambiguation is overly generous, and I suspect we can figure out a more optimized way to traverse those character lists. Let's chat through some issues in office hours.